src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,183 @@
+        {
+            var session = (WebTransportSession)state!;
+            session.OnClientConnectionClosed();
+        }, this);
+    }
+
+    void IWebTransportSession.Abort(int errorCode)
+    {
+        Abort(new(), (Http3ErrorCode)errorCode);
+    }
+
+    internal void OnClientConnectionClosed()
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        _connectionClosedRegistration.Dispose();
+
+        lock (_openStreams)
+        {
+            foreach (var stream in _openStreams)
+            {
+                _ = stream.Value.DisposeAsync().AsTask();
+            }
+
+            _openStreams.Clear();
+        }
+
+        _pendingStreams.Writer.Complete();
+    }
+
+    internal void Abort(ConnectionAbortedException exception, Http3ErrorCode error)
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        _connectionClosedRegistration.Dispose();
+
+        lock (_openStreams)
+        {
+            _connectStream.Abort(exception, error);
+            foreach (var stream in _openStreams)
+            {
+                if (exception.InnerException is not null)
+                {
+                    stream.Value.Abort(new ConnectionAbortedException(exception.Message, exception.InnerException));
+                }
+                else
+                {
+                    stream.Value.Abort(new ConnectionAbortedException(exception.Message));
+                }
+            }
+            _openStreams.Clear();
+        }
+
+        _pendingStreams.Writer.Complete();
+    }
+
+    public async ValueTask<ConnectionContext?> OpenUnidirectionalStreamAsync(CancellationToken cancellationToken)
+    {
+        if (_isClosing)
+        {
+            return null;
+        }
+        // create the stream
+        var features = new FeatureCollection();
+        features.Set(_outputStreamDirectionFeature);
+        var connectionContext = await _connection._multiplexedContext.ConnectAsync(features, cancellationToken);
+        var streamContext = _connection.CreateHttpStreamContext(connectionContext);
+        var stream = new WebTransportStream(streamContext, WebTransportStreamType.Output);
+
+        var success = _openStreams.TryAdd(stream.StreamId, stream);
+        Debug.Assert(success);
+
+        // send the stream header
+        // https://ietf-wg-webtrans.github.io/draft-ietf-webtrans-http3/draft-ietf-webtrans-http3.html#name-unidirectional-streams
+        await stream.Transport.Output.WriteAsync(OutputStreamHeader, cancellationToken);
+
+        return stream;
+    }
+
+    internal void AddStream(WebTransportStream stream)
+    {
+        if (_isClosing)
+        {
+            throw new ObjectDisposedException(CoreStrings.WebTransportIsClosing);
+        }
+
+        if (!_pendingStreams.Writer.TryWrite(stream) || !_openStreams.TryAdd(stream.StreamId, stream))
```

Comment:
It should be tracked locally before it's given to the application. TryAdd also seems less likley to fail.
```suggestion
        if (!_openStreams.TryAdd(stream.StreamId, stream) || !_pendingStreams.Writer.TryWrite(stream))
```